File: src/utils/updater.ts
/**
 * GitHub 更新检查与外部链接打开（关于页共用）
 *
 * - 版本比较：纯函数，便于单测；网络请求封装为可注入的 fetchLatestRelease。
 * - 打开链接：Tauri 环境走 shell plugin，浏览器环境回退 window.open。
 */

import { isTauri } from '@/api/tauri'
import { invoke } from '@/api'
import { errorMessage } from '@/utils/error'

export const GITHUB_REPO = 'meng-plus/serial-tools'
export const GITHUB_REPO_URL = `https://github.com/${GITHUB_REPO}`
export const GITHUB_ISSUES_URL = `${GITHUB_REPO_URL}/issues`
export const GITHUB_RELEASES_URL = `${GITHUB_REPO_URL}/releases`
export const GITHUB_RELEASES_API = `https://api.github.com/repos/${GITHUB_REPO}/releases/latest`

/** GitHub Release 字段子集（/releases/latest 响应） */
export interface GitHubRelease {
  tag_name: string
  name: string
  html_url: string
  published_at: string
  body: string | null
}

/** 解析版本号（去 v 前缀、去 prerelease 后缀），取 major/minor/patch */
export function parseVersion(v: string): [number, number, number] {
  const cleaned = v.replace(/^[vV]/, '').trim()
  const parts = cleaned.split('-')[0].split('.')
  return [
    Number(parts[0]) || 0,
    Number(parts[1]) || 0,
    Number(parts[2]) || 0,
  ]
}

/**
 * 比较两个版本号：a > b 返回 1，相等返回 0，a < b 返回 -1
 * 主 / 次 / 补丁逐段比较
 */
export function compareVersions(a: string, b: string): number {
  const pa = parseVersion(a)
  const pb = parseVersion(b)
  for (let i = 0; i < 3; i++) {
    if (pa[i] !== pb[i]) return pa[i] > pb[i] ? 1 : -1
  }
  return 0
}

/**
 * 拉取 GitHub 最新正式版；网络 / 解析失败抛错。
 * Tauri 环境走后端命令（规避 WebView fetch 的 CORS/UA 环境差异，错误分类更精确）；
 * 浏览器预览回退直接 fetch。
 */
export async function fetchLatestRelease(): Promise<GitHubRelease> {
  if (isTauri()) {
    try {
      return await invoke<GitHubRelease>('check_for_update')
    } catch (e) {
      throw new Error(errorMessage(e))
    }
  }
  const res = await fetch(GITHUB_RELEASES_API, {
    headers: { Accept: 'application/vnd.github+json' },
    cache: 'no-store',
  })
  if (!res.ok) {
    if (res.status === 403 && res.headers.get('x-ratelimit-remaining') === '0') {
      throw new Error('GitHub API 请求频率超限，请稍后重试或前往 GitHub 查看')
    }
    throw new Error(`GitHub API 返回 ${res.status}`)
  }
  return (await res.json()) as GitHubRelease
}

export interface UpdateCheckResult {
  hasUpdate: boolean
  latest: GitHubRelease | null
  error: string | null
}

/**
 * 对比当前版本与 GitHub 最新正式版。
 * 网络 / 解析失败不抛出，而是返回 error 供 UI 提示。
 */
export async function checkForUpdate(currentVersion: string): Promise<UpdateCheckResult> {
  try {
    const latest = await fetchLatestRelease()
    if (compareVersions(latest.tag_name, currentVersion) > 0) {
      return { hasUpdate: true, latest, error: null }
    }
    return { hasUpdate: false, latest: null, error: null }
  } catch (e) {
    return { hasUpdate: false, latest: null, error: e instanceof Error ? e.message : String(e) }
  }
}

/** 打开外部链接：Tauri 用 shell plugin，浏览器用 window.open */
export async function openExternal(url: string): Promise<void> {
  if (isTauri()) {
    const { open } = await import('@tauri-apps/plugin-shell')
    await open(url)
  } else {
    window.open(url, '_blank', 'noopener')
  }
}

File: src-tauri/src/commands/updater.rs
//! GitHub 最新版本检查 — 网络请求在 Rust 后端完成，规避 WebView fetch 的
//! CORS / UA 环境差异，并对 403 限流做分类提示。

use crate::error::CommandError;
use std::time::Duration;

/// GitHub Release 信息（/releases/latest 响应子集，序列化键与前端契约一致）
#[derive(Debug, serde::Serialize)]
pub struct UpdateInfo {
    pub tag_name: String,
    pub name: String,
    pub html_url: String,
    pub published_at: String,
    pub body: Option<String>,
}

const GITHUB_RELEASES_API: &str =
    "https://api.github.com/repos/meng-plus/serial-tools/releases/latest";

/// 将 HTTP 错误分类为可操作的中文提示（纯函数，便于单测）。
///
/// - 403 且限流剩余为 0：明确为频率超限
/// - 其余 403：多为网络代理 / IP 信誉拦截
/// - 404：仓库尚无 release
fn classify_http_error(status: u16, ratelimit_remaining: Option<&str>) -> String {
    match status {
        403 if ratelimit_remaining == Some("0") => {
            "GitHub API 请求频率超限（60 次/小时），请稍后重试或前往 GitHub 查看".to_string()
        }
        403 => "GitHub API 拒绝请求（可能被网络代理或限流拦截），可前往 GitHub 查看".to_string(),
        404 => "GitHub 仓库暂无正式发布版本".to_string(),
        s => format!("GitHub API 返回 {s}，可前往 GitHub 查看"),
    }
}

/// 拉取最新正式版 release 信息
#[tauri::command]
pub async fn check_for_update() -> Result<UpdateInfo, CommandError> {
    let agent = ureq::AgentBuilder::new()
        .timeout(Duration::from_secs(10))
        .build();
    let version = env!("CARGO_PKG_VERSION");

    let result = agent
        .get(GITHUB_RELEASES_API)
        .set("Accept", "application/vnd.github+json")
        .set(
            "User-Agent",
            &format!("serial-tools-update-check/{version}"),
        )
        .call();

    match result {
        Ok(resp) => {
            let value: serde_json::Value = serde_json::from_reader(resp.into_reader())
                .map_err(|e| CommandError::Internal(format!("解析 GitHub 响应失败: {e}")))?;
            let into = |v: &serde_json::Value| v.as_str().unwrap_or_default().to_string();
            let body = value
                .get("body")
                .and_then(|b| b.as_str())
                .map(str::to_string);
            Ok(UpdateInfo {
                tag_name: into(value.get("tag_name").unwrap_or(&serde_json::Value::Null)),
                name: into(value.get("name").unwrap_or(&serde_json::Value::Null)),
                html_url: into(value.get("html_url").unwrap_or(&serde_json::Value::Null)),
                published_at: into(
                    value
                        .get("published_at")
                        .unwrap_or(&serde_json::Value::Null),
                ),
                body,
            })
        }
        Err(ureq::Error::Status(code, resp)) => {
            let remaining = resp.header("X-RateLimit-Remaining").map(str::to_string);
            let _ = resp.into_string();
            Err(CommandError::Internal(classify_http_error(
                code,
                remaining.as_deref(),
            )))
        }
        Err(e) => Err(CommandError::Internal(format!("无法连接 GitHub API: {e}"))),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn classify_rate_limited() {
        assert_eq!(
            classify_http_error(403, Some("0")),
            "GitHub API 请求频率超限（60 次/小时），请稍后重试或前往 GitHub 查看"
        );
    }

    #[test]
    fn classify_forbidden_not_rate_limited() {
        let msg = classify_http_error(403, Some("57"));
        assert!(msg.contains("拒绝请求"));
    }

    #[test]
    fn classify_other_status() {
        assert!(classify_http_error(404, None).contains("暂无正式发布"));
        assert!(classify_http_error(500, None).contains("GitHub API 返回 500"));
    }
}
